### PR TITLE
Fix backup poller races

### LINKED ISSUE #25959: Core dump in gpr_mu_unlock
### What version of gRPC and what language are you using?

grpc 1.18 c++

### What operating system (Linux, Windows,...) and version?
Freebsd 11

### What runtime / compiler are you using (e.g. python version or version of gcc)
llvm3.7

### What did you do?
Please provide either 1) A unit test for reproducing the bug or 2) Specific steps for us to follow to reproduce the bug. If there’s not enough information to debug the problem, gRPC team may close the issue at their discretion. You’re welcome to re-open the issue once you have a reproduction.
We have a server streaming RPC that is designed to stream response to the clients periodically.  On one of the device, a crash was seen during the streaming. We also noticed network congestion prior to the crash. Similar issue was reported earlier: https://github.com/grpc/grpc/issues/25273
Based on grpc-dev feedback, we had patched our stack with changes from #20139.

This issue is tough to reproduce but is seen with the 1.18.0 version of stack. Is there a known issue in this area and if so can there be references given to any fixes that we could try out. Else any suggestions to identify the root cause or to reliably reproduce the issue would be helpful.

### What did you expect to see?
No crash.

### What did you see instead?

Core with below bt

Core seen during long run with bt:
#0  thr_kill () at thr_kill.S:3

#1  0x08568745 in abort () at /volume/build/junos/occam/llvm-3.7/sandbox-20200416/freebsd/stable_11/20200416.020549_builder_stable_11.d6acdc5/src/lib/libc/stdlib/abort.c:65

#2  0x082ca25b in gpr_mu_unlock (mu=0xb6298ac) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/gpr/sync_posix.cc:90

#3  0x082f17ee in cover_self (tcp=<optimized out>) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/iomgr/tcp_posix.cc:245

#4  notify_on_write (tcp=0x9a20000) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/iomgr/tcp_posix.cc:270

#5  0x082f1a17 in tcp_handle_write (arg=0x9a20000, error=<optimized out>) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/iomgr/tcp_posix.cc:941

#6  0x082bf964 in exec_ctx_run (closure=0x9a20108, error=0x0) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/iomgr/exec_ctx.cc:40

#7  0x082bf871 in grpc_core::ExecCtx::Flush (this=0xfe1ee8e0) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/iomgr/exec_ctx.cc:147

#8  0x082a732b in grpc_core::ExecCtx::~ExecCtx (this=0x85c5c44 <vtable for grpc_core::ExecCtx+8>) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/iomgr/exec_ctx.h:108

#9  grpc_call_start_batch (call=<optimized out>, ops=<optimized out>, nops=<optimized out>, tag=<optimized out>, reserved=<optimized out>) at ../../../../../../../src/external/apache2/grpc/dist/src/core/lib/surface/call.cc:1912

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

### PR DESCRIPTION
**Pull request by yashykt**
Fix #25959

Currently, the refcounting around `g_backup_poller`, does not make sure that the value from `g_backup_poller` is the most current value.

Consider the following scenario - 
1) A backup poller `x` is in use and has a refcount (`g_uncovered_notifications_pending`) of 1 (implying that it is no longer needed).
2) Thread 1 invokes `run_poller` notices the refcount of 1, decrements it to 0.
3) Thread 2 invokes `cover_self`, adds 2 refs (0->2), notices that the count was previously 0 and proceeds to creating a new backup poller `y`.
4) Thread 3 also invokes `cover_self`, adds 2 refs (2->4), but when it reads the value of g_backup_poller, it still reads `x` since `g_backup_poller` hasn't yet been updated. This could lead to crashes as seen in #25959

This is just one sequence of events proving the bugginess of the code. There are potentially other similar sequences that can cause issues. The fix seems to be that updates to `g_backup_poller` and `g_uncovered_notifications_pending` should happen atomically which is easiest handled by a mutex.

### CODE REVIEW

**Review by drfloob — COMMENTED** [review]
This looks like it'd work out. It's great to see more use of ABSL_GUARDED_BY! I did have some thoughts about introducing asserts here, but they're really minor - we'd be replacing one type of crash with another if there are other race conditions we haven't seen yet. But ... I'm not seeing them.

Were you able to reproduce the problem in N*1000 runs, by any chance?

Also, I'd like one more set of eyes on this to be sure.

File: src/core/lib/iomgr/tcp_posix.cc [hunk 1]
```diff
@@ -489,13 +488,16 @@ static void run_poller(void* bp, grpc_error_handle /*error_ignored*/) {
...
+  old_count = g_uncovered_notifications_pending--;
+  g_backup_poller_mu->Unlock();
+  GPR_ASSERT(old_count > 1);
   if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
-    gpr_log(GPR_INFO, "BACKUP_POLLER:%p uncover cnt %d->%d", p,
-            static_cast<int>(old_count), static_cast<int>(old_count) - 1);
+    gpr_log(GPR_INFO, "BACKUP_POLLER:%p uncover cnt %d->%d", p, old_count,
+            old_count - 1);
   }
   GPR_ASSERT(old_count != 1);
```
[the remark this answers is not in the document]
Comment (yashykt) [reply src/core/lib/iomgr/tcp_posix.cc]: Done

**Review by yashykt — COMMENTED** [review]
Haven't been able to reproduce this, nor do I remember seeing this failure anywhere other than #25959

**Comment by nicolasnoble** [comment]
This looks good to me at face value - we may want to give it a try in the uv-ee branch and see what happens.

**Comment by yashykt** [comment]
I was under the impression that uv-ee would not use tcp_posix.cc...

**Comment by nicolasnoble** [comment]
Sorry, got my wires crossed a bit here. Yes, I see what you're doing now. LGTM :)

**Review by nicolasnoble — APPROVED** [review]

**Comment by yashykt** [comment]
Thanks for reviewing!

**Pull request merged**